src/openms/source/METADATA/SpectrumLookup.cpp

```diff
@@ -0,0 +1,397 @@
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Hendrik Weisser $
+// $Authors: Hendrik Weisser $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/METADATA/SpectrumLookup.h>
+
+#include <OpenMS/CONCEPT/LogStream.h>
+#include <OpenMS/FORMAT/FileHandler.h>
+
+using namespace std;
+
+namespace OpenMS
+{
+  const String& SpectrumLookup::regexp_names_ = "INDEX0 INDEX1 SCAN ID RT";
+
+  SpectrumLookup::SpectrumLookup(): 
+    rt_tolerance(0.01), spectra_(0), n_spectra_(0),
+    regexp_name_list_(ListUtils::create<String>(regexp_names_, ' '))
+  {}
+
+
+  SpectrumLookup::~SpectrumLookup()
+  {}
+
+
+  bool SpectrumLookup::empty() const
+  {
+    return n_spectra_ == 0;
+  }
+  
+
+  void SpectrumLookup::setSpectra(vector<MSSpectrum<> >& spectra, 
+                                  const String& scan_regexp)
+  {
+    spectra_ = &spectra;
+    n_spectra_ = spectra.size();
+    rts_.clear();
+    ids_.clear();
+    scans_.clear();
+    boost::regex re;
+    if (!scan_regexp.empty())
+    {
+      if (!scan_regexp.hasSubstring("?<SCAN>"))
+      {
+        String msg = "The regular expression for extracting scan numbers from native IDs must contain a named group '?<SCAN>'.";
+        throw Exception::IllegalArgument(__FILE__, __LINE__,
+                                         __PRETTY_FUNCTION__, msg);
+      }
+      re.assign(scan_regexp);
+    }
+    for (Size i = 0; i < n_spectra_; ++i)
+    {
+      MSSpectrum<>& spec = spectra[i];
+      String native_id = spec.getNativeID();
+      rts_[spec.getRT()] = i;
+      ids_[native_id] = i;
+      if (!scan_regexp.empty())
+      {
+        boost::smatch match;
+        bool found = boost::regex_search(native_id, match, re);
+        if (found && match["SCAN"].matched)
+        {
+          String value = match["SCAN"].str();
+          try
+          {
+            Size scan_no = value.toInt();
+            scans_[scan_no] = i;
+            continue;
+          }
+          catch(Exception::ConversionError&)
+          {
+          }
+        }
+        LOG_WARN << "Warning: Could not extract scan number from spectrum native ID '" + native_id + "' using regular expression '" + scan_regexp + "'. Look-up by scan number may not work properly." << endl;
+      }
+    }
+  }
+
+  
+  MSSpectrum<>& SpectrumLookup::findByRT(double rt) const
+  {
+    double upper_diff = numeric_limits<double>::infinity();
+    map<double, Size>::const_iterator upper = rts_.upper_bound(rt);
+    if (upper != rts_.end())
+    {
+      upper_diff = upper->first - rt;
+    }
+    double lower_diff = numeric_limits<double>::infinity();
+    map<double, Size>::const_iterator lower = upper;
+    if (lower != rts_.begin())
+    {
+      --lower;
+      lower_diff = rt - lower->first;
+    }
+    if ((lower_diff < upper_diff) && (lower_diff <= rt_tolerance))
+    {
+      return (*spectra_)[lower->second];
```

Comment:
That's what this code should do. Given an RT value x, I first find the closest spectrum with RT <= x and calculate the RT difference. Next I find the previous (in RT) spectrum and also calculate the difference. Then I choose the spectrum with the smaller difference.

Or is there a bug that I'm not seeing?